sdk/core/src/policy.rs

```diff
@@ -118,30 +68,68 @@ impl Context {
 }
 
 #[async_trait::async_trait]
-pub trait Policy: Send + Sync + std::fmt::Debug {
-    async fn send(
-        &self,
-        ctx: Context,
-        request: &mut Request,
-        next: &[Arc<dyn Policy>],
-    ) -> PolicyResult<Response>;
+pub trait Policy<R>: Send + Sync + std::fmt::Debug
+where
+    R: Send + Sync,
+{
+    async fn send(&self, ctx: Context, request: Request) -> PolicyResult<R>;
 }
 
-#[derive(Debug, Clone)]
+#[derive(Clone)]
 pub struct Pipeline {
-    policies: Vec<Arc<dyn Policy>>,
+    policies: Vec<Arc<dyn Policy<Request>>>,
+    retry: Arc<dyn RetryPolicy>,
+    transport: Arc<dyn Policy<Response>>,
 }
 
 impl Pipeline {
-    // TODO: how can we ensure that the transport policy is the last policy?
-    // Make this more idiot proof
-    pub fn new(policies: Vec<Arc<dyn Policy>>) -> Self {
-        Self { policies }
+    pub fn new(
+        policies: Vec<Arc<dyn Policy<Request>>>,
+        retry: Arc<dyn RetryPolicy>,
+        transport: Arc<dyn Policy<Response>>,
+    ) -> Self {
+        Self {
+            policies,
+            retry,
+            transport,
+        }
     }
 
-    pub async fn send(&self, ctx: Context, mut request: Request) -> PolicyResult<Response> {
-        self.policies[0]
-            .send(ctx, &mut request, &self.policies[1..])
-            .await
+    pub async fn send(&self, ctx: Context, request: Request) -> PolicyResult<Response> {
+        // with create instance we make sure every pipeline execution has a new retry policy
+        // instance (with for example the counters reset to zero) based off the template policy
+        // specified in the pipeline.
+        let mut retry_policy = self.retry.create_instance().await;
```

Comment:
I think we should keep this as a RetryPolicy. Policies are supposed to be fully customizable even to the point of removing or rescheduling where this retry is performed. Making everything a policy right up to and including the TransportPolicy has been working well for all the other SDKs.